### ISSUE TITLE
SearchBar fails to call onSubmitted Function

### ISSUE THREAD

**Issue by MuhammedOzdogan** [flutter/flutter#137210]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

- Copy the example usage of [SearchBar](https://api.flutter.dev/flutter/material/SearchBar-class.html) from official Flutter documentation.
- Add [onSubmitted](https://api.flutter.dev/flutter/material/SearchBar/onSubmitted.html) property to SearchBar object it can be as simple as `onSubmitted: (v) => print("Search value: $v"),`
- Run the application
- Click the Search Bar
- Type `something` and hit enter
- `onSubmitted` callback doesn't' get called.

`onSubmitted` never getting called, if SearchBar suggestions are visible, [contoller.openView()](https://api.flutter.dev/flutter/material/SearchController/openView.html)

### Expected results

I expect to `onSubmitted` callback being called whenever user press `ENTER` or `DONE` button in software keyboard.

### Actual results

`onSubmitted` doesn't get called.

### Code sample

https://dartpad.dev/?id=9474848a10119912f0efd690546c8c8d

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

#### Only way to make it work is closing the SuggestionsWidget and then hitting ENTER. 

https://github.com/flutter/flutter/assets/26257939/cb6c5312-5a31-41e9-8777-0fe46b52cde4

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
➜  bug flutter doctor -v
[✓] Flutter (Channel stable, 3.13.8, on macOS 14.0 23A344 darwin-arm64, locale en-TR)
    • Flutter version 3.13.8 on channel stable at /Users/muhammed/apps/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6c4930c4ac (7 days ago), 2023-10-18 10:57:55 -0500
    • Engine revision 767d8c75e8
    • Dart version 3.1.4
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/muhammed/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.83.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.74.0```

</details>

**Comment by dam-ease** [comment]
Hi @MuhammedOzdogan. Thanks for filing this. Running this on my end, this is working as expected. From the docs,

> Called when the user indicates that they are done editing the text in the field.

This seems to occur after clicking DONE when in the `SearchBar` view.

https://github.com/flutter/flutter/assets/53122008/9563323a-5820-41b8-9241-6cc8b0c85197

Can you provide more information on this if I'm missing something on your issue?

<details><summary>Code Sample</summary>
<p>

```dart
import 'package:flutter/material.dart';

/// Flutter code sample for [SearchBar].

void main() => runApp(const SearchBarApp());

class SearchBarApp extends StatefulWidget {
  const SearchBarApp({super.key});

  @override
  State<SearchBarApp> createState() => _SearchBarAppState();
}

class _SearchBarAppState extends State<SearchBarApp> {
  bool isDark = false;

  @override
  Widget build(BuildContext context) {
    final ThemeData themeData = ThemeData(
        useMaterial3: true,
        brightness: isDark ? Brightness.dark : Brightness.light);

    return MaterialApp(
      theme: themeData,
      home: Scaffold(
        appBar: AppBar(title: const Text('Search Bar Sample')),
        body: Padding(
          padding: const EdgeInsets.all(8.0),
          child: SearchAnchor(
              builder: (BuildContext context, SearchController controller) {
            return SearchBar(
              controller: controller,
              padding: const MaterialStatePropertyAll<EdgeInsets>(
                  EdgeInsets.symmetric(horizontal: 16.0)),
              onTap: () {
                controller.openView();
              },
              onChanged: (_) {
                controller.openView();
              },
              onSubmitted: (v) => print("Search value: $v"),
              leading: const Icon(Icons.search),
              trailing: <Widget>[
                Tooltip(
                  message: 'Change brightness mode',
                  child: IconButton(
                    isSelected: isDark,
                    onPressed: () {
                      setState(() {
                        isDark = !isDark;
                      });
                    },
                    icon: const Icon(Icons.wb_sunny_outlined),
                    selectedIcon: const Icon(Icons.brightness_2_outlined),
                  ),
                )
              ],
            );
          }, suggestionsBuilder:
                  (BuildContext context, SearchController controller) {
            return List<ListTile>.generate(5, (int index) {
              final String item = 'item $index';
              return ListTile(
                title: Text(item),
                onTap: () {
                  setState(() {
                    controller.closeView(item);
                  });
                },
              );
            });
          }),
        ),
      ),
    );
  }
}

``` 

</p>
</details> 

<details><summary>flutter doctor -v</summary>
<p>

```
[✓] Flutter (Channel stable, 3.13.8, on macOS 14.0 23A344 darwin-arm64, locale
    en-NG)
    • Flutter version 3.13.8 on channel stable at
      /Users/damilolaalimi/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6c4930c4ac (7 days ago), 2023-10-18 10:57:55 -0500
    • Engine revision 767d8c75e8
    • Dart version 3.1.4
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/damilolaalimi/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/damilolaalimi/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)

[!] Android Studio (version unknown)
    • Android Studio at /Users/damilolaalimi/Downloads/Android Studio
      Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.83.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] VS Code (version 1.83.1)
    • VS Code at /Users/damilolaalimi/Downloads/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • iPhone 15 Pro Max (mobile) • 56A225C9-B8F6-4201-A0AE-8C3CFA4CE152 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-17-0 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-arm64   • macOS 14.0 23A344 darwin-arm64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 118.0.5993.117

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
``` 

</p>
</details>

**Comment by MuhammedOzdogan** [comment]
Hello @dam-ease,

You are picking up a value from the Suggestions (`item 0`, `item 1`...), type something doesn't exist and hit ENTER/DONE.
For example type `test` and hit enter.

**Comment by dam-ease** [comment]
Thanks for your response. Not sure entering a text that doesn't belong to the `suggestionBuilder` should trigger submission in the `SerachBar`.
After typing, `test` for instance, the search through `suggestion` `TextInput` field doesn't get dismissed for the submission to be triggered. 
This seems to be working as expected or can you share more on why `submission` should be triggered when done typing a text different from items in `suggestionBuilder`?

**Comment by MuhammedOzdogan** [comment]
Entering a text that doesn't belong to the suggestionBuilder should definitely trigger a submission.
As it's name implies it's just *suggestion* not *options*, it suggest something I may or may not pick it up.
It's as simple as that. Just open your Gmail application, type something there are **suggestions** but it's allow you to **search** something not suggested.

> This seems to be working as expected or can you share more on why submission should be triggered when done typing a text different from items in suggestionBuilder?

Because this is not a Selectbox or [Dropdown](https://api.flutter.dev/flutter/material/DropdownMenu-class.html) Menu, I'm not using it to give the user options to pick up. I'm using SearchBar to search something.

**Comment by dam-ease** [comment]
Thanks for explaining further. I'll keep this open and label it for more insights.

**Comment by QuncCccccc** [comment]
Thanks a lot for filing this issue! I think there is a similar issue #132915. The fix is in #136840. I'll close this as duplicated.
